driver/src/main/java/eu/cloudnetservice/driver/network/rpc/listener/RPCPacketListener.java

```diff
@@ -60,173 +59,218 @@ public RPCPacketListener(@NonNull RPCHandlerRegistry rpcHandlerRegistry) {
-          } else {
-            // an exception was thrown previously, break
-            buf.readString(); // remove the handler information which is not necessary
-            result = this.serializeResult(
-              lastResult,
-              lastResult.invocationHandler().dataBufFactory(),
-              lastResult.invocationHandler().objectMapper(),
-              this.buildContext(channel, buf, null, true));
-            break;
-          }
-        } else {
-          // just process over to remove the content from the buffer
-          this.handleRaw(buf.readString(), this.buildContext(channel, buf, null, true));
+    var content = packet.content();
+    var resultExpected = packet.uniqueId() != null;
+
+    try {
+      var rpcDepth = content.readInt();
+      if (rpcDepth <= 0) {
+        // depth must be at least one (single call) or more (chained call)
+        if (resultExpected) {
+          var resultContent = DataBufFactory.defaultFactory()
+            .createWithExpectedSize(1)
+            .writeByte(RPCInvocationResult.STATUS_BAD_REQUEST)
+            .writeString("invalid chain length");
+          this.sendResponseData(channel, packet, resultContent);
         }
+        return;
       }
-      // check if there is already a result (which is caused by an exception - we can skip the handling step then)
-      if (result == null && lastResult != null) {
-        // the last handler decides over the method invocation result
-        result = this.handle(
-          buf.readString(),
-          this.buildContext(channel, buf, lastResult.invocationResult(), true));
+
+      if (rpcDepth > 1) {
+        // RPC chain, start executing the first step
+        this.executeRPCChainStep(rpcDepth, 1, resultExpected, content, packet, channel, null);
+      } else {
+        // single method rpc, execute & respond if requested
+        var targetClassName = content.readString();
+        var invocationContext = this.buildContext(content, null);
+        var handlingTask = this.postRPCRequestToHandler(targetClassName, invocationContext);
+        if (resultExpected) {
+          this.waitForInvocationCompletion(handlingTask, result -> {
+            var resultContent = this.serializeHandlingResult(result);
+            this.sendResponseData(channel, packet, resultContent);
+          });
+        }
       }
-    } else {
-      // just invoke the method
-      result = this.handle(buf.readString(), this.buildContext(channel, buf, null, false));
-    }
-    // check if we need to send a result
-    if (result != null && packet.uniqueId() != null) {
-      var response = new BasePacket(-1, result);
-      response.uniqueId(packet.uniqueId());
-      channel.sendPacket(response);
+    } finally {
+      // specifically release the buffer here to prevent memory leaks, especially if we didn't consume
+      // the whole buffer content (for example due to an exception during handling)
+      content.release();
```

Comment:
Is this maybe a place for a forceRelease?

Reply:
Yea, we could. The assumption is: due to this being a packet handler the packet buffer shouldn't be retained multiple times when the handler is called, therefore there shouldn't be a difference between release and forceRelease.